### PR TITLE
Rewrite SetextHeaderSyntax

### PR DESCRIPTION
**Pull request by chenzhiguang**
The performance is also much better according to the new algorithm. 

### Before

Screenshot 2022-12-01 at 06 40 02

### After

Screenshot 2022-12-01 at 06 39 05

### CODE REVIEW

**Comment by chenzhiguang** [comment]
The flutter_markdown test will be green when this PR is merged: [Refactor list syntax](https://github.com/dart-lang/markdown/pull/499)

**Comment by chenzhiguang** [comment]
Ready for review!

File: lib/src/block_parser.dart [hunk 1]
```diff
@@ -114,10 +126,21 @@ class BlockParser {
   BlockSyntax? get parentSyntax => _parentSyntax;
   BlockSyntax? _parentSyntax;
 
+  /// If the [SetextHeadingSyntax] is disabled temporarily
```
Comment (srawlins) [inline lib/src/block_parser.dart:129]: Start this comment with "Whether", and end with a period.

https://dart.dev/guides/language/effective-dart/documentation#prefer-starting-a-boolean-variable-or-property-comment-with-whether-followed-by-a-noun-or-gerund-phrase

File: lib/src/block_parser.dart [hunk 2]
```diff
@@ -114,10 +126,21 @@ class BlockParser {
...
+  bool get setextHeadingDisabled => _setextHeadingDisabled;
+  bool _setextHeadingDisabled = false;
+
+  /// The [BlockSyntax] which is running now.
```
Comment (srawlins) [inline lib/src/block_parser.dart:133]: Could you note in this comment when `currentSyntax` is `null`.

File: lib/src/block_syntaxes/blockquote_syntax.dart [hunk 3]
```diff
@@ -19,10 +19,15 @@ class BlockquoteSyntax extends BlockSyntax {
 
   const BlockquoteSyntax();
 
+  /// If this blockquote is ended up at a lazy continuation line.
```
Comment (srawlins) [inline lib/src/block_syntaxes/blockquote_syntax.dart:22]: I think this should start with "Whether" [1] and maybe say "ends with a lazy continuation line." Does that sound right?

[1] https://dart.dev/guides/language/effective-dart/documentation#prefer-starting-a-boolean-variable-or-property-comment-with-whether-followed-by-a-noun-or-gerund-phrase

File: lib/src/block_syntaxes/blockquote_syntax.dart [hunk 4]
```diff
@@ -19,10 +19,15 @@ class BlockquoteSyntax extends BlockSyntax {
...
+  // The definition of  lazy continuation lines:
```
Comment (srawlins) [inline lib/src/block_syntaxes/blockquote_syntax.dart:23]: extra space before the word 'lazy'.

File: lib/src/block_parser.dart [hunk 5]
```diff
@@ -35,6 +35,18 @@ class BlockParser {
...
 
+  /// Starting line of the last unconsumed content.
+  int _start = 0;
+
+  /// The lines from [_start] to [_pos] (inclusive), it works as a buffer for
+  /// some blocks, for example:
+  /// When the [ParagraphSyntax] parsing process is interrupted by the
+  /// [SetextHeaderSyntax], so this structure is not a paragraph but a setext
+  /// heading, then the [ParagraphSyntax] does not have to retreat the reading
+  /// position, it only needs to return `null`, the [SetextHeaderSyntax] will
```
Comment (srawlins) [inline lib/src/block_parser.dart:46]: Let's mention the method name here, something like "then the [ParagraphSyntax.parse] method does not..."

**Review by srawlins — CHANGES_REQUESTED** [review]
Looks great, just few comments. Thanks again!

File: lib/src/block_parser.dart (hunk 1 above)
Comment (chenzhiguang) [reply lib/src/block_parser.dart]: Fixed! Thanks!

File: lib/src/block_parser.dart (hunk 5 above)
Comment (chenzhiguang) [reply lib/src/block_parser.dart]: Cool, we support this [A.B] format now.

File: lib/src/block_syntaxes/blockquote_syntax.dart (hunk 4 above)
Comment (chenzhiguang) [reply lib/src/block_syntaxes/blockquote_syntax.dart]: Fixed

File: lib/src/block_syntaxes/blockquote_syntax.dart (hunk 3 above)
Comment (chenzhiguang) [reply lib/src/block_syntaxes/blockquote_syntax.dart]: Thanks. Fixed!

File: lib/src/block_parser.dart (hunk 2 above)
Comment (chenzhiguang) [reply lib/src/block_parser.dart:133]: Fixed

**Comment by chenzhiguang** [comment]
@srawlins Hi, can we continue with this one?

**Review by srawlins — APPROVED** [review]
Love it, thanks!

**Pull request merged**